### ISSUE TITLE
DM timeline loads (/api/v1/timelines/direct) are unusually slow

### ISSUE THREAD

**Issue by rtucker** [glitch-soc/mastodon#397]
I'm noticing that loading the contents of the direct message column is unusually slow for me, like on the order of 15+ seconds.  It looks to be mostly database delay, which suggests that an index might be missing:

web_1 | [fe97475a-b47d-4bf4-b73a-f1cacbbecf86] method=GET path=/api/v1/timelines/direct format=html controller=Api::V1::Timelines::DirectController action=show status=200 duration=15092.58 view=1619.73 db=13042.11

Haven't investigated much yet, just wanted to get this written down so I can forget about it ^^

vulpineclub issue: https://github.com/vulpineclub/mastodon/issues/4

* * * *

- [X] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by yipdw** [comment]
Here's what `Status.as_direct_timeline` does on my instance (glitchsoc at around tootsuite v2.3.1):

```
irb(main):016:0> Status.as_direct_timeline(Account.find(1)).explain
=> EXPLAIN for: SELECT "statuses".* FROM "statuses" LEFT OUTER JOIN mentions ON statuses.id = mentions.status_id AND mentions.account_id = 1 LEFT OUTER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE (mentions.account_id = 1 OR statuses.account_id = 1) AND "statuses"."visibility" = 3 AND ("statuses"."account_id" NOT IN ([ids])) AND "accounts"."silenced" = $1 ORDER BY "statuses"."id" DESC [["silenced", "f"]]
                                                                      QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=37182.05..37182.06 rows=4 width=443)
   Sort Key: statuses.id DESC
   ->  Nested Loop  (cost=1.14..37182.01 rows=4 width=443)
         ->  Nested Loop Left Join  (cost=0.85..37162.58 rows=4 width=443)
               Filter: ((mentions.account_id = 1) OR (statuses.account_id = 1))
               ->  Index Scan using index_statuses_20180106 on statuses  (cost=0.42..34305.99 rows=359 width=443)
                     Index Cond: (visibility = 3)
                     Filter: (account_id <> ALL ('{[ids]}'::bigint[]))
               ->  Index Scan using index_mentions_on_status_id on mentions  (cost=0.42..7.94 rows=1 width=16)
                     Index Cond: (statuses.id = status_id)
                     Filter: (account_id = 1)
         ->  Index Scan using accounts_pkey on accounts  (cost=0.29..4.85 rows=1 width=8)
               Index Cond: (id = statuses.account_id)
               Filter: (NOT silenced)
(14 rows)
```

The `[ids]` bit is account IDs that were added by timeline filters.

The index scan on `index_statuses_20180106` is going to be very expensive (and the cost reflects that): it's an index scan on _all statuses in the database_, excluding a relatively small number of account IDs.

Maybe we can improve this by starting at a smaller dataset: say, starting from `mentions` where `account_id` is equal to the account in question?  This would miss DMs initiated by said account; we'll have to use some other mechanism to pull those in.

**Comment by rtucker** [comment]
Here's my explain:

```
irb(main):001:0> Status.as_direct_timeline(Account.find(1)).explain
=> EXPLAIN for: SELECT "statuses".* FROM "statuses" LEFT OUTER JOIN mentions ON statuses.id = mentions.status_id AND mentions.account_id = 1 LEFT OUTER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE (mentions.account_id = 1 OR statuses.account_id = 1) AND "statuses"."visibility" = 3 AND ("statuses"."account_id" NOT IN ([ids])) AND "accounts"."silenced" = $1 ORDER BY "statuses"."id" DESC [["silenced", "f"]]
                                                                                                                                                       QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sort  (cost=224541.03..224541.15 rows=47 width=441)
   Sort Key: statuses.id DESC
   ->  Nested Loop  (cost=206450.25..224539.73 rows=47 width=441)
         ->  Hash Right Join  (cost=206449.96..224390.13 rows=48 width=441)
               Hash Cond: (mentions.status_id = statuses.id)
               Filter: ((mentions.account_id = 1) OR (statuses.account_id = 1))
               ->  Bitmap Heap Scan on mentions  (cost=343.87..18251.21 rows=8702 width=16)
                     Recheck Cond: (account_id = 1)
                     ->  Bitmap Index Scan on index_mentions_on_account_id_and_status_id  (cost=0.00..341.69 rows=8702 width=0)
                           Index Cond: (account_id = 1)
               ->  Hash  (cost=206017.86..206017.86 rows=7058 width=441)
                     ->  Index Scan using index_statuses_20180106 on statuses  (cost=0.56..206017.86 rows=7058 width=441)
                           Index Cond: (visibility = 3)
                           Filter: (account_id <> ALL ('{[ids]}'::bigint[]))
         ->  Index Scan using index_accounts_on_id on accounts  (cost=0.29..3.11 rows=1 width=8)
               Index Cond: (id = statuses.account_id)
               Filter: (NOT silenced)
(17 rows)
```

So yep, definitely costly :<

Some sage advice from statuses_controller.rb:

```
  def account_media_status_ids
    # `SELECT DISTINCT id, updated_at` is too slow, so pluck ids at first, and then select id, updated_at with ids.
    # Also, Avoid getting slow by not narrowing down by `statuses.account_id`.
    # When narrowing down by `statuses.account_id`, `index_statuses_20180106` will be used
    # and the table will be joined by `Merge Semi Join`, so the query will be slow.
    Status.joins(:media_attachments).merge(@account.media_attachments).permitted_for(@account, current_account)
          .paginate_by_max_id(limit_param(DEFAULT_STATUSES_LIMIT), params[:max_id], params[:since_id])
          .reorder(id: :desc).distinct(:id).pluck(:id)
  end
```

Pondering intensifies...

**Comment by yipdw** [comment]
There was some work on this in https://github.com/tootsuite/mastodon/commit/b87a1229c74aebec02cd08091f25613f6dff4428; this was merged into glitch-soc on 2018-05-30 as part of 165b5dc7f5711efd07adb73316e94d89b9e7e3a3.  Can you try the DM tim

**Comment by yipdw** [comment]
There was some work on this in https://github.com/tootsuite/mastodon/commit/b87a1229c74aebec02cd08091f25613f6dff4428; this was merged into glitch-soc on 2018-05-30 as part of 165b5dc7f5711efd07adb73316e94d89b9e7e3a3.  Can you update and try the DM timeline again, see if it's any faster?

**Comment by rtucker** [comment]
I have notifications, local timeline, three lists, DM, and Home as columns, in that order, so invalidate-and-reload gets pretty busy and seems to be a pretty good performance test.  All tests are in production on vulpine.club, as myself.

I tested with a build from **before** 165b5dc and time-to-first-byte was in the 1.9 second range, which is on par with the other timelines.  So, I think there's been some other work.

Just to be sure, I tested against f009fef just now and TTFB is still right around 1.9 seconds.

I'm happy.  Closing :)
